src/User/User.php

```diff
@@ -320,13 +333,28 @@ public function getDisplayNameAttribute()
      */
     public function checkPassword($password)
     {
+        /**
+         * @deprecated beta 14, remove in beta 15.
+         */
         $valid = static::$dispatcher->until(new CheckingPassword($this, $password));
 
         if ($valid !== null) {
             return $valid;
         }
 
-        return static::$hasher->check($password, $this->password);
+        $valid = false;
+
+        foreach (static::$passwordCheckers as $checker) {
+            $result = $checker($this, $password);
+
+            if ($result === false) {
+                return false;
+            } elseif ($result === true) {
+                $valid = true;
```

Comment:
What's the benefit of not immediately returning false here. In any future iteration the value will only be set to true if true anyway. Allowing more iteration will only cost us performance here.

Reply:
Do you mean returning true? It's because an output of false overrides any outputs of true, regardless of order of evaluation. Unless order is a feature (as with middleware), I'm trying to maximally avoid order of evaluation mattering. 

As such, an explicit false, denies, an explicit true allows as long as there are no explicit denials, and no trues means deny.

Reply:
So if I understand correctly, we need at least one return to be `=== true` and none `=== false` for `checkPassword` to return `true`.

I'm assuming it's so checks can return `null` without impacting the result if they have neither success nor failure.

Reply:
To clarify, checkers should return as follows:

- If the password is valid, return true
- If not, return null
- If the request must be blocked, return false

This is what's currently used in Clark's extensions. Before I merge this in, is this schema fine with everyone?

Reply:
Maybe we could offer CONST aliases just like we did for policies? I'm ok with the raw value, but maybe it would be a bit easier in the documentation if each of these option had a human readable constant associated with it?

Reply:
Hmm, I'm not sure. They'd have to be properties of the User object. More importantly though, I have no idea what to call the `null` one. `User::PASSWORD_VALID` and `User::PASSWORD_INVALID` make sense, but maybe `User::PASSWORD_SKIP`?

Reply:
Not sure :upside_down_face: 

Maybe I'd call them something including AUTHORIZE, FAIL_AUTHORIZATION and something with PASS/SKIP/NEXT?

Reply:
Also, technically you don't really need to return `null` ever. It's also what happens when you don't return anything so there might be a need for a constant.

Reply:
I think that this might make things more confusing to be completely honest. Let's keep it as is for now; if we see confusion in the developer community we can always add aliases later I think?

> Also, technically you don't really need to return null ever. It's also what happens when you don't return anything so there might not be a need for a constant.

True, I'll try to clarify this in the docblock